main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1069 @@
+        // - have better performance (we don't generate unnecessary type variables)
+        //
+        val knownTarget = exp match {
+          case KindedAst.Expr.Def(sym, tvar2, loc2) =>
+            // Case 1: Lookup the sym and instantiate its scheme.
+            val defn = root.defs(sym)
+            val (tconstrs0, declaredType) = Scheme.instantiate(defn.spec.sc, loc2.asSynthetic)
+            val tconstrs = tconstrs0.map(_.copy(loc = loc))
+            Some((sym, tvar2, tconstrs, declaredType))
+
+          case KindedAst.Expr.Sig(sym, tvar2, loc2) =>
+            // Case 2: Lookup the sym and instantiate its scheme.
+            val sig = root.classes(sym.clazz).sigs(sym)
+            val (tconstrs0, declaredType) = Scheme.instantiate(sig.spec.sc, loc2.asSynthetic)
+            val tconstrs = tconstrs0.map(_.copy(loc = loc))
+            Some((sym, tvar2, tconstrs, declaredType))
+
+          case _ =>
+            // Case 3: Unknown target.
+            None
+        }
+
+        knownTarget match {
+          case Some((sym, tvar2, constrs1, declaredType)) =>
+            //
+            // Special Case: We are applying a Def or Sig and we break apart its declared type.
+            //
+            val declaredEff = declaredType.arrowEffectType
+            val declaredArgumentTypes = declaredType.arrowArgTypes
+            val declaredResultType = declaredType.arrowResultType
+
+            val (tpes, effs) = exps.map(visitExp).unzip
+            c.expectTypeArguments(sym, declaredArgumentTypes, tpes, exps.map(_.loc), loc)
+            c.addClassConstraintsM(constrs1, loc)
+            c.unifyTypeM(tvar2, declaredType, loc)
+            c.unifyTypeM(tvar, declaredResultType, loc)
+            c.unifyTypeM(evar, Type.mkUnion(declaredEff :: effs, loc), loc)
+            val resTpe = tvar
+            val resEff = evar
+            (resTpe, resEff)
+
+          case None =>
+            //
+            // Default Case: Apply.
+            //
+            val lambdaBodyType = Type.freshVar(Kind.Star, loc)
+            val lambdaBodyEff = Type.freshVar(Kind.Eff, loc)
+            val (tpe, eff) = visitExp(exp)
+            val (tpes, effs) = exps.map(visitExp).unzip
+            c.expectTypeM(tpe, Type.mkUncurriedArrowWithEffect(tpes, lambdaBodyEff, lambdaBodyType, loc), loc)
+            c.unifyTypeM(tvar, lambdaBodyType, loc)
+            c.unifyTypeM(evar, Type.mkUnion(lambdaBodyEff :: eff :: effs, loc), loc)
+            val resTpe = tvar
+            val resEff = evar
+            (resTpe, resEff)
+        }
+
+      case Expr.Lambda(fparam, exp, loc) =>
+        c.unifyTypeM(fparam.sym.tvar, fparam.tpe, loc)
+        val (tpe, eff) = visitExp(exp)
+        val resTpe = Type.mkArrowWithEffect(fparam.tpe, eff, tpe, loc)
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case KindedAst.Expr.Unary(sop, exp, tvar, loc) => sop match {
+        case SemanticOp.BoolOp.Not =>
```

Comment:
It seems all of these unary operations could use the same `helper(exp, tvar, type)` function

Reply:
listed